splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/JoinOperation.java

```diff
@@ -208,10 +207,17 @@ public void writeExternal(ObjectOutput out) throws IOException {
 						out.writeObject(rightRow);
                         out.writeObject(mergedRowTemplate);
 				}
-				boolean sparkExpressionPresent = (sparkJoinPredicate != null);
+				boolean sparkExpressionPresent = (sparkExpressionTreeAsString != null &&
+				                                  sparkExpressionTreeAsString.length() != 0);
+
 				out.writeBoolean(sparkExpressionPresent);
-				if (sparkExpressionPresent)
-					out.writeObject(sparkJoinPredicate);
+				if (sparkExpressionPresent) {
```

Comment:
Would it be possible to delay the decoding further to when it's actually used? For NLJoins we serialize the operation in order to clone it, so we'd be paying the overhead anyway.

Reply:
@dgomezferro Better yet, we don't even need to build the SparkExpressionNode tree unless the join is broadcast join or mergesort join.  I pushed a change to this effect.  Joins on control other than nested loop join are not serialized I believe, so there should be no need to delay deserialization any further with the join strategy check in place.

Reply:
There might still be some performance impact because when we have a NLJ we serialize/deserialize the whole operation tree, so a broadcast join on the same tree would get serialized. I think it's ok for now and we can revisit later if NLJ becomes expensive

Reply:
I see.  I pushed a change to delay deserialization of the predicates tree until it's used.  Up until then if we need to serialize/deserialize, we'll do so on the string representation of the tree.